### ISSUE TITLE
handle packet old_buf_addr being truncated

### ISSUE THREAD

**Issue by LauraLMann** [panda-re/panda#504]
It appears some of the changes in pull request #497 have broken the old_buf_addr argument to the panda_callbacks_replay_handle_packet function.  It used to be a uint64_t, but now it's a target_ptr_t.  When building i386 on a 64-bit host, target_ptr_t is uint32_t.  If you look at rr_log.c line 833, it appears old_buf_addr is being set to a host address, which on a 64-bit host is a 64-bit number.  But because the old_buf_addr is now 32-bits in all those functions it is passed too (and C/C++ is a really lazy language when it comes to type checking), we quietly lose the top 4 bytes of old_buf_addr.
As old_buf_addr is valuable when trying to coordinate packet processing during replays, I really think we need to keep the original value, and revert to having old_buf_addr be the size of a HOST pointer.

**Comment by nathanjackson** [comment]
Indeed. The `old_buf_addr` type is `uint64_t` in the non-deterministic log.

https://github.com/panda-re/panda/blob/master/panda/include/panda/rr/rr_log.h#L102

This callback needs to be changed back.

**Comment by m000** [comment]
Indeed I may have gotten this wrong. The argument is described in the callbacks definition header as `XXX this is a mystery`, so I went with the wrong guess, thinking the argument was referring to a guest address.

Since this is cast to `uintptr_t` in `rr_log.c`, using that instead of `target_ptr_t` _should_ work. I'll update #501 in a minute to also include a fix for this.

**Comment by m000** [comment]
Looking closer, this is a bit more complicated. In the rare usage scenario where someone attempts to replay on a 32bit host a trace captured on a 64bit host, the cast on `rr_log.c:833` will result in truncation. So, there's that.

**Comment by m000** [comment]
@LauraLMann Can you help me clarify the use of `old_buf_addr`? I'm getting conflicting input from the code.

- As you mention, in `rr_log.c:833`, the value is loaded from the rr log as if it refers to a host address.
- In `manual.md:1898`, `old_buf_addr` is described as "the address that `buf` had when the recording was taken". So again it sounds like a host address.
- But in tainted net plugin (`tainted_net.cpp:122`), the address is used to apply taint. I.e. it is used as a guest address.

Based on the use, I'd say that `target_ptr_t` is the correct type. I'm not sure what kind of loss we have for the 4 top bytes of `old_buf_add`. For a 32bit guest, these should be 0 anyway.

**Comment by nathanjackson** [comment]
The IO shadow is basically a shadow of QEMU's internal buffers for devices. The `taint2_label_io` call is not labeling RAM or CPU. Rather it is shadowing the emulated device buffers and registers.

When a packet comes in, it hits an internal buffer inside PANDA (this would be like a FIFO in the real hardware). Then the packet gets copied into the guest some how by the running kernel (IO ports, MMIO, DMA, etc). `tainted_net`'s `taint2_label_io` call is labeling packets while they're still sitting in a device side buffer before it has even been copied into the CPU or RAM.

Even devices are emulated by the host, so the emulated buffers are basically sitting in host memory. Therefore `old_buf_addr` must be a host address.

**Comment by LauraLMann** [comment]
Thanks, @nathanjackson , that was my vague recollection too of why old_buf_addr is really a host address even though it is used to update a taint shadow of the emulated network buffer.  The shadow isn't of something in the guest, but of a buffer in the emulation of the guest device.

**Comment by m000** [comment]
@nathanjackson So if I understand correctly, `taint2_label_io` is used to extend tainting to emulated device buffers?

**Comment by nathanjackson** [comment]
That's my understanding.

**Comment by m000** [comment]
Ok, I've opened a new PR. I've opted for `uint64_t`. You can tentatively use this.

However, I suspect that the value of `old_buf_addr` is truncated when one replays traces captured on a 64bit host on a 32bit host. I'll try to verify that, and maybe improve the PR.

**Comment by m000** [comment]
#505 should now be good to merge.

**Comment by LauraLMann** [comment]
It ends up that #505 was only a partial fix to the problem, as not all of the type changes were reverted properly in the tainted_net plugin.
Also, the network transfer callbacks (PANDA_CB_REPLAY_NET_TRANSFER) have a similar problem in the source and destination addresses - sometimes one or the other or even both of those addresses are host addresses too, and thus need to be uint64_t on a 64-bit host or they get truncated after being read from the recording.
As I already figured out how to to fix all this as a part of debugging newly failing regression tests, I'm going to submit a pull request shortly.  This is just a heads up to @m000 and whoever else care that it is coming.
